File: 833-find-and-replace-in-string/go/findReplaceString.go
type Item struct {
	Length int
	Target string
}

func findReplaceString(s string, indices []int, sources []string, targets []string) string {
	replaceMap, n, m := make(map[int]Item), len(indices), len(s)
	var (
		index    int
		source   string
		item     string
		endIndex int
	)
	for i := 0; i < n; i++ {
		index = indices[i]
		source = sources[i]
		endIndex = index + len(source)
		if endIndex > m {
			endIndex = m
		}
		item = string(s[index:endIndex])
		if item == source {
			replaceMap[index] = Item{len(source), targets[i]}
		}
	}
	result, i := "", 0

	for i < m {
		if temp, ok := replaceMap[i]; ok {
			result += temp.Target
			i += temp.Length
		} else {
			result += string(s[i])
			i++
		}
	}

	return result
}

File: 833-find-and-replace-in-string/js/findReplaceString.js
/**
 * @param {string} s
 * @param {number[]} indices
 * @param {string[]} sources
 * @param {string[]} targets
 * @return {string}
 */
var findReplaceString = function (s, indices, sources, targets) {
  // find all that need replace
  // 寻找所有需要被替换的
  let map = new Map();
  const n = indices.length;
  let index, source, item;

  for (let i = 0; i < n; i++) {
    index = indices[i];
    source = sources[i];
    item = s.substr(index, source.length);

    if (item === source) {
      map.set(index, [item.length, targets[i]]);
    }
  }
  // do replace
  const result = [];
  const m = s.length;
  let i = 0;
  while (i < m) {
    item = map.get(i);
    if (item) {
      result.push(item[1]);
      i += item[0];
    } else {
      result.push(s[i]);
      i++;
    }
  }

  return result.join("");
};
